server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/ExtractMDNOriginalJMAPMessageId.java

```diff
@@ -0,0 +1,162 @@
+        try {
+            mimeMessage.addHeader(X_JAMES_MDN_JMAP_MESSAGE_ID, messageId.serialize());
+        } catch (MessagingException e) {
+            LOGGER.error("unable to add " + X_JAMES_MDN_JMAP_MESSAGE_ID + " header to message", e);
+        }
+    }
+
+    private Optional<MessageId> findMessageIdForRFC822MessageId(String messageId, MailAddress recipient) {
+        try {
+            MailboxSession session = mailboxManager.createSystemSession(recipient.asString());
+            int limit = 1;
+            MultimailboxesSearchQuery searchByRFC822MessageId = MultimailboxesSearchQuery
+                .from(new SearchQuery(SearchQuery.mimeMessageID(messageId)))
+                .build();
+            return mailboxManager.search(searchByRFC822MessageId, session, limit).stream().findFirst();
+        } catch (MailboxException e) {
+            LOGGER.error("unable to find message with Message-Id: " + messageId, e);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<MDNReport> parseReport(Entity report) {
+        try {
+            return new MDNReportParser().parse(((SingleBody)report.getBody()).getInputStream(), report.getCharset());
+        } catch (IOException e) {
+            LOGGER.error("unable to parse MESSAGE_DISPOSITION_NOTIFICATION part", e);
+            return Optional.empty();
+        }
+    }
+
+    private Optional<Entity> findReport(MimeMessage mimeMessage) {
+        return parseMessage(mimeMessage).flatMap(this::extractReport);
+    }
+
+    @VisibleForTesting Optional<Entity> extractReport(Message message) {
+        if (!message.isMultipart()) {
+            LOGGER.debug("MDN Message must be multipart");
+            return Optional.empty();
+        }
+        List<Entity> bodyParts = ((Multipart) message.getBody()).getBodyParts();
+        if (bodyParts.size() < 2) {
+            LOGGER.debug("MDN Message must contain at least two parts");
+            return Optional.empty();
+        }
+        Entity report = bodyParts.get(1);
+        if (!isDispositionNotification(report)) {
+            LOGGER.debug("MDN Message second part must be of type " + MESSAGE_DISPOSITION_NOTIFICATION);
+            return Optional.empty();
+        }
+        return Optional.of(report);
+    }
+
+    private boolean isDispositionNotification(Entity entity) {
+        return entity
+            .getMimeType()
+            .startsWith(MESSAGE_DISPOSITION_NOTIFICATION);
+    }
+
+    private Optional<Message> parseMessage(MimeMessage mimeMessage) {
+        try {
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            mimeMessage.writeTo(os);
+            Message message = new DefaultMessageBuilder().parseMessage(new ByteArrayInputStream(os.toByteArray()));
+            return Optional.of(message);
+        } catch (IOException | MessagingException e) {
+            LOGGER.error("unable to parse message", e);
+        }
+        return Optional.empty();
```

Comment:
I'd prefer that optional to be in the catch block. Then it is clear it is what happens in the face of errors.